Yarn documentation for Cachito in README.md

akhmanova opened the pull request:
There is finally the first version of documentation for yarn. Sorry for keeping it long, I have some kind of fear of English texts. 

There are changes for Nexus part and a new section for package managers category.

akhmanova commented on README.md:
```diff
@@ -589,16 +591,16 @@ Nexus instance that hosts the permanent content: `cachito_nexus_hoster_username`
+Baseline                | ✓     | ✓   | ✓   | ✓    |
+Content Manifest        | ✓     | ✓   | ✓   | ✓    |
+Dependency Replacements | ✓     | x   | x   | x    |
+Dev Dependencies        | ✓     | ✓   | ✓   | x    |
```
Should we count the ability to keep dev dependencies as ` ✓ `? I did not do it because yarn can not distinguish dev deps.

chmeliik replied:
Let's keep it as `x` until we add the ability to do that :+1:

chmeliik replied:
Speaking of dev dependencies, why is gomod marked as supported??

athos-ribeiro replied:
For Go, we divide them into two categories: module level dependencies (whatever is needed to build the package) and package level dependencies (what was actually imported by the modules. This does not include all the packages provided by the imported modules).

There is a misconception here though, since they are not dev dependencies (dependencies required for testing would be included in both since we do not know what the user will actually build and ship).

We could keep it as checket with a `*`, `✓*`, or `✓¹`, and then add a note under the table to make this one clear. WDYT?

chmeliik replied:
Sounds good (doesn't need to be in this PR though).

akhmanova commented on README.md:
```diff
@@ -530,17 +530,19 @@ with the `mod_auth_gssapi` module.
 ## Nexus
 
-### Nexus For npm
+### Nexus For Java Script
```
I decided to have one section for both yarn and npm because yarn is just a CNAME for js nexus repos.

chmeliik commented on README.md:
```diff
@@ -530,17 +530,19 @@ with the `mod_auth_gssapi` module.
+[Nexus Repository Manager 3][nexus-docs] to store JS dependencies. The Nexus instance will have a
+JS group repository (e.g. `cachito-js`) which points to a JS hosted repository (e.g.
+`cachito-js-hosted`) and a JS proxy repository
+(e.g. `cachito-js-proxy`) that points to registry.npmjs.org or registry.yarnpkg.com according to
```
`cachito-js-proxy` points to registry.npmjs.org for both npm and yarn. The reason why this works is that the 2 hostnames actually point to the same server (i.e. registry.yarnpkg.com is a CNAME for registry.npmjs.org).

athos-ribeiro replied:
This. And it would be nice to mention it here, if w eare going to describe this process in such detailed level.

chmeliik commented on README.md:
```diff
@@ -799,3 +801,42 @@ curl "localhost:8080/api/v1/requests" \
 
+### yarn
+
+Cachito handles the yarn package manager in much the same way as the [npm](###npm) package manager.
```
```suggestion
Cachito handles the yarn package manager in much the same way as the [npm](#npm) package manager.
```
See the difference between
https://github.com/akhmanova/cachito/tree/yarn-documentation###npm
https://github.com/akhmanova/cachito/tree/yarn-documentation#npm

chmeliik commented on README.md:
```diff
@@ -799,3 +801,42 @@ curl "localhost:8080/api/v1/requests" \
+All requests for the yarn package manager with `package-lock.json`, `npm-shrinkwrap.json` files in
+the root directory will fail because those files are dedicated for [npm](###npm).
+
+After parsing Cachito creates a yarn registry in an instance of Nexus it manages that contains just
```
```suggestion
After parsing, Cachito creates a yarn registry in an instance of Nexus it manages that contains just
```

chmeliik commented on README.md:
```diff
@@ -799,3 +801,42 @@ curl "localhost:8080/api/v1/requests" \
+dependencies can be added. The connection information is stored in an
+[.npmrc](https://docs.npmjs.com/configuring-npm/npmrc.html) file accessible at the
+`/api/v1/requests/<id>/configuration-files` API endpoint. There will be an empty
+[.yarnrc](https://yarnpkg.com/configuration/yarnrc) file in the same directory with the
```
This link is for the Yarn 2 config file, the old one (which we use) is https://classic.yarnpkg.com/en/docs/yarnrc

chmeliik commented on README.md:
```diff
@@ -799,3 +801,42 @@ curl "localhost:8080/api/v1/requests" \
+will fail the request. When Cachito encounters a supported location, it will download the
+dependency, modify the version in the [package.json](https://docs.npmjs.com/files/package.json) to
+be unique, upload it to Nexus, modify the top level project's
+[package.json](https://docs.npmjs.com/files/package.json) and lock files to use the dependency from
```
```suggestion
[package.json](https://docs.npmjs.com/files/package.json) and yarn.lock to use the dependency from
```
You can also include a link if you want

chmeliik reviewed (commented):
Nice! I have a few nitpicks, but overall seems good :-)

athos-ribeiro commented on README.md:
```diff
@@ -410,15 +410,15 @@ Custom configuration for the Celery workers are listed below:
   the Nexus instance that has the hosted repositories. This is used instead of
   `cachito_nexus_password` for uploading content if you are using the two Nexus instance approach as
-  described in the "Nexus For npm" section. If this is set, `cachito_nexus_hoster_username` must
+  described in the "Nexus For Java Script" section. If this is set, `cachito_nexus_hoster_username` must
```
On a side note, the description of two nexus approaches should be moved to a general section since it is and will be used for more package managers.

athos-ribeiro commented on README.md:
```diff
@@ -589,16 +591,16 @@ Nexus instance that hosts the permanent content: `cachito_nexus_hoster_username`
+External Dependencies   | N/A   | ✓   | ✓   | ✓    |
+Multiple Paths          | ✓     | ✓   | ✓   | ✓    |
+Nested Dependencies     | ✓     | ✓   | x   | ✓    |
+Offline Installations   | ✓     | x   | x   | x    |
```
It would be nice to introduce a supported version field here (this could happen in a separate PR).
We know we do not support pip < 10, and we also know we only support Yarn versions >= 1.12.0, <  2. We also only support Go versions from when modules were fully introduced (I believe it was 1.13). Finally we also know there are limitations on npm >= 7 support (we do not allow yarn.lock file parsing for npm).

chmeliik replied:
Now that I see the field in place, I'm wondering if this isn't more confusing than helpful. For many of these version constraints, the issue is an edge case which will only be encountered by some users. For example:
- pip < 10 only breaks if you use non-PyPI deps
- npm >= 7 only breaks if you use yarn.lock as the lockfile
- yarn < 1.12 only breaks if you use https deps
- yarn >= 1.12 also breaks if you added the https dep with an older version

IMO we should document these exceptions in the sections dedicated to the specific package managers. Speaking of exceptions, even if we do not include the supported versions field, it would be good to mention that yarn 2 is not supported.

akhmanova replied:
@chmeliik Do you want me to create a tech improvement story for the future with it?

chmeliik replied:
+1, also note that the pip point is already documented: https://github.com/release-engineering/cachito/blob/master/docs/pip.md

chmeliik commented on README.md:
```diff
@@ -583,22 +572,37 @@ In this case, you will need to configure the following additional settings that
 Nexus instance that hosts the permanent content: `cachito_nexus_hoster_username`,
 `cachito_nexus_hoster_password`, and `cachito_nexus_hoster_url`.
 
+On each request, Cachito will create a proxy repository to the group repository of the request's
+package manager (e.g. `cachito-pkg-manager`). Cachito will populate this proxy repository to
+contain the subset of dependencies declared in the repository's lock file. Once populated, Cachito
+will block the repository from getting additional content. This prevents the consumer of the
+repository from installing something that was not declared in the lock file. This is further
+enforced by locking down the repository to a single user created for the request, which the
+consumer will use. Please keep in mind that for this to function properly, anonymous access needs
+to be disabled on the Nexus instance or at least not set to have read access on all repositories.
+
+These repositories and users created per request are deleted when the request is marked as stale
+or the request fails.
```
I don't think this should be in the general section, many of these points are unique to npm and yarn at the moment.

chmeliik commented on README.md:
```diff
@@ -530,27 +530,16 @@ with the `mod_auth_gssapi` module.
+[Nexus Repository Manager 3][nexus-docs] to store JS dependencies. The Nexus instance will have a
+JS group repository (e.g. `cachito-js`) which points to a JS hosted repository (e.g.
+`cachito-js-hosted`) and a JS proxy repository
+(e.g. `cachito-js-proxy`) that points to the npm/yarn registry (registry.npmjs.org, registry.yarnpkg.com).
```
Can we also mention that registry.npmjs.org = registry.yarnpkg.com?

chmeliik commented on README.md:
```diff
@@ -583,22 +572,37 @@ In this case, you will need to configure the following additional settings that
+Nested Dependencies     |  ✓      | ✓   | x     | ✓              |
+Offline Installations   |  ✓      | x   | x     | x              |
+
+[^1] Cachito keeps Dev Dependencies for yarn packages but can not distinguish them as it done for npm.
```
I don't think we need extra explanation for this. IMO this is what `x` means.

It *could* be misunderstood as "yarn does not download dev dependencies at all," but if we wanted this table to be precise, we would have more exceptions than rules.

Maybe the explanation for Dev Dependencies should clearly state that it only applies to Cachito's ability to determine if something is a dev dependency or not.

chmeliik replied:
I believe Athos's point about adding clarification was for gomod, which is reasonable but does not need to be part of this PR I think.

athos-ribeiro commented on README.md:
```diff
@@ -583,22 +572,37 @@ In this case, you will need to configure the following additional settings that
-Offline Installations   | ✓     | x   | x   |
+Feature                 |  gomod  | npm | pip   | yarn           |
+---                     |  ---    | --- | ---   | ---            |
+Versions supported      | >= 1.13 | < 7 | >= 10 | >= 1.12.0, < 2 |
```
Thanks! :)

akhmanova commented:
@athos-ribeiro @chmeliik 

Changes after your last review: 
1. Deleted ✓^1 symbol and comment from yarn column
2. Moved nexus js part back from nexus common configuration.
3. added info that both npm and yarn registries point to the same server

Actions for me after the PR will be merged: 
1. Create a story to add information about supported versions of package managers (keep in mind that information about pip)
2. Create a pr with ✓^1 symbol  and comment for gomod devDependencies support.

chmeliik reviewed (approved):
Nice!

One technical detail: the commits do not have a Signed-off-by line, you can add that by running `git commit --amend -s` (not sure if that's required here, it is required in osbs)

akhmanova commented:
Ok, I just added a sign for both commits, thank you, @chmeliik !
